providers/implementations/encode_decode/decode_der2key.c

```diff
@@ -206,34 +206,51 @@ static int der2key_decode(void *vctx, OSSL_CORE_BIO *cin, int selection,
 
     ok = 0;                      /* Assume that we fail */
 
+    ERR_set_mark();
     if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
         derp = der;
         if (ctx->desc->d2i_PKCS8 != NULL) {
             key = ctx->desc->d2i_PKCS8(NULL, &derp, der_len, ctx);
-            if (ctx->flag_fatal)
+            if (ctx->flag_fatal) {
+                ERR_clear_last_mark();
                 goto end;
+            }
         } else if (ctx->desc->d2i_private_key != NULL) {
             key = ctx->desc->d2i_private_key(NULL, &derp, der_len);
         }
-        if (key == NULL && ctx->selection != 0)
+        if (key == NULL && ctx->selection != 0) {
+            ERR_clear_last_mark();
             goto next;
+        }
     }
     if (key == NULL && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
+        ERR_pop_to_mark();
+        ERR_set_mark();
```

Comment:
I am unsure about this and the same two calls below. I understand that in some cases this might mean there will be more errors added than necessary but adding these calls also might hide real errors from the previous decoding attempts. Let's say the thing to decode is a malformed encoding of a private key then this will remove the errors from the private key decoding attempt.

Reply:
Execution can enter line 227 only if `selection` enables parsing both a private key and a public key (typically because the user did not specify what he expects) and a non-fatal error occurred trying to parse a private key.
Good point that the two calls can lead to wrongly discarding error queue entries resulting form that.
So removed the calls here and below.